taskcat/stacker.py

```diff
@@ -1722,8 +1722,92 @@ def check_cfnyaml(self, yamlin, quiet=None, strict=None):
+        templates = {}
+        config = yaml.safe_load(open(self.config))
+        rules = cfnlint.core.get_rules([], [])
+        for test in config['tests'].keys():
+            lints[test] = {}
+            if 'regions' in config['tests'][test].keys():
+                lints[test]['regions'] = config['tests'][test]['regions']
+            else:
+                lints[test]['regions'] = config['global']['regions']
+            template_file = config['tests'][test]['template_file']
+            template_file = 'templates/' + template_file
+            lints[test]['template_file'] = template_file
+            if template_file not in templates.keys():
+                templates[template_file] = self.get_child_templates(template_file)
+            lints[test]['results'] = {}
+            templates[template_file].add(template_file)
+            for t in templates[template_file]:
+                template = cfnlint.decode.cfn_yaml.load(t)
+                lints[test]['results'][t] = cfnlint.core.run_checks(
+                    t,
+                    template,
+                    rules,
+                    lints[test]['regions'])
+        test_status = {"W": False, "E": False}
+        for test in lints.keys():
+            for t in lints[test]['results'].keys():
+                print(I + "Lint results for test %s on template %s:" % (test, t))
+                for r in lints[test]['results'][t]:
+                    message = r.__str__().lstrip('[')
+                    sev = message[0]
+                    if sev == 'E':
+                        print(E + "    " + message)
+                        test_status[sev] = True
+                    elif sev == 'W':
+                        if 'E' + message[1:] not in [r.__str__().lstrip('[') for r in lints[test]['results'][t]]:
+                            print(I + "    " + message)
+                            test_status[sev] = True
+                    else:
+                        print(D + "linter produced unkown output: " + message)
+        if strict in ['error', 'strict'] and test_status['E']:
+            print(F + "Exiting due to lint errors")
+            sys.exit(1)
+        elif strict == 'strict' and test_status['W']:
+            print(F + "Exiting due to lint warnings")
+            sys.exit(1)
 
+    def get_child_templates(self, filename):
+        """
+        recursively find nested templates given a template path
 
+        :param filename: string, path to template
+        :return: set of nested template paths
+        """
+        children = set()
+        template = cfnlint.decode.cfn_yaml.load(filename)
+        for resource in template['Resources'].keys():
+            child_name = ''
+            if template['Resources'][resource]['Type'] == "AWS::CloudFormation::Stack":
+                template_url = template['Resources'][resource]['Properties']['TemplateURL']
+                if type(template_url) == dict:
+                    if 'Fn::Sub' in template_url.keys():
```

Comment:
You're only accounting for 'Fn::Sub', not '!Sub'.

Reply:
the (awesome) linting library that we're now using to parse templates converts all shorthand for us. so !Sub becomes Fn::Sub by the time this code gets to it

Reply:
👍